man/man6/wsolitaire.6
.TH WSOLITAIRE 6 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wsolitaire \- classic Solitaire board game for W
.SH SYNOPSIS
wsolitaire [-u]
.SH DESCRIPTION
.I Wsolitaire
is a one player board game.  The board consists of pieces and 'holes' in
which the pieces are.  At the beginning only the hole in the middle is empty.
You can remove piece from the board by 'jumping' over it with another piece
(to an empty hole). Your mission is to remove all the pieces from the board
except the last one, which you should try to get into the middle hole at the
end.
.PP
You can undo the last move with the right mouse button. By using the '-u'
command line option, you'll have unlimited UNDOs. However, that's
cheating...
.PP
You can exit the program at the middle of a game with the 'q' key.
.SH LAST CHANGED
12/1995 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


games/wsolitaire.c
/*
 * wsolitaire.c, a part of the W Window System
 *
 * Copyright (C) 1995 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- Solitaire (a well known 'remove pieces from board') game
 */

#include <stdio.h>		/* fprintf() */
#include <Wlib.h>


/* W specific stuff */

#define WIN_PROPERTIES	(W_MOVE | W_TITLE | W_CLOSE | EV_MOUSE | EV_KEYS)
#define WIN_NAME	" SOLITAIRE "

/* board and pieces are square (latter actually circle) sized */
#define BLOCK	13			/* block size (should be odd) */
#define LARGE	5			/* radiuses for the circles inside */
#define SMALL	3
#define SIZE	9			/* board size */
#define PIECES	44			/* pieces on the board at start */

/* what can be at a board position */
#define SPACE	0
#define EMPTY	1
#define PIECE	2

/* user actions defines */
#define EXIT	0
#define PROCEED	1
#define UNDO	2

#define DIST(a,b)	((b) > (a) ? (b) - (a) : (a) - (b))


/* global variables */
static WWIN *Window;

static unsigned char Board[SIZE][SIZE];
static short FromX, FromY;


/* prototypes */
static int  initialize(char *argv[]);
static void restore(int signal);
static void init_board(void);
static void draw_piece(short x, short y);
static void select_piece(short x, short y);
static int  move_from(short x, short y);
static int  move_to(short x, short y);
static void push(short x, short y);
static int  undo(void);
static int  wait_input(void);


int main(int argc, char *argv[])
{
  int
    action,
    done_already = 1, enabled = 0,	/* 1 / unlimited undo */
    todo = PIECES, valid = 0;		/* moves to do / valid move */
  short
    mx, my;

  /* check for unlimited undo argument */
  if(argc > 1 && argv[1][0] == '-' && argv[1][1] == 'u' && argv[1][2] == '\0')
    enabled = 1;

  /* initialize W */
  if(!initialize(argv))
  {
    restore(1);
    return 1;
  }

  w_setmode(Window, M_INVERS);
  init_board();

  do
  {
    if((action = wait_input()) == EXIT)
      break;

    if(action == UNDO)
    {
      if(valid)
        select_piece(FromX, FromY);		/* deselect */
      valid = 0;

      if(enabled || !done_already)
      {
        done_already = 1;
        if(undo())
          todo++;
      }
      continue;
    }

    w_querymousepos(Window, &mx, &my);
    mx /= BLOCK;
    my /= BLOCK;

    if(valid)
    {
      if(move_to(mx, my))		/* with checks of course */
      {
        done_already = 0;
        push(mx, my);			/* store for undo-pop */
        todo--;
      }
      valid = 0;			/* next move */
    }
    else
      valid = move_from(mx, my);	/* valid moe beginning? */
  } while(todo > 1);

  if(todo == 1)
  {
    wait_input();
    /* congratulate! */
  }

  restore(0);
  return 0;
}

/* --------------------
 * Initialize W stuff (connect server, open window, set font/title)
 * and allocate Board. Return 0 on error.
 */
static int initialize(char *argv[])
{
  WSERVER *wserver;

  if(!(wserver = w_init()))
  {
    fprintf(stderr, "%s: can't connect to wserver\n", *argv);
    return 0;
  }

  if(!(Window = w_create(BLOCK * SIZE, BLOCK * SIZE, WIN_PROPERTIES)))
  {
    fprintf(stderr, "%s: can't create window\n", *argv);
    return 0;
  }
  if(w_open(Window, UNDEF, UNDEF) < 0)
  {
    fprintf(stderr, "%s: can't open window\n", *argv);
    return 0;
  }
  w_settitle(Window, WIN_NAME);
  return 1;
}

/* signal / exit handler */
static void restore(int signal)
{
  if(Window)
  {
    w_exit();
  }
}


/* initialize and draw the Board. */
static void init_board(void)
{
  short x, y, xx, yy;

  for(y = 0; y < SIZE; y++)
    for(x = 0; x < SIZE; x++)
      Board[x][y] = SPACE;

  for(y = 0; y < SIZE; y++)
    for(x = 3; x < SIZE-3; x++)
      Board[x][y] = PIECE;

  for(y = 3; y < SIZE-3; y++)
    for(x = 0; x < SIZE; x++)
      Board[x][y] = PIECE;

  Board[SIZE / 2][SIZE / 2] = EMPTY;

  yy = BLOCK / 2;
  for(y = 0;   y < SIZE; y++)
  {
    xx = BLOCK / 2;
    for(x = 0; x < SIZE; x++)
    {
      switch(Board[x][y])
      {
	case PIECE:
	  w_pcircle(Window, xx, yy, SMALL);
	case EMPTY:
	  w_circle(Window, xx, yy, LARGE);
      }
      xx += BLOCK;
    }
    yy += BLOCK;
  }
}

/* ----------------------- */
static void draw_piece(short x, short y)
{
  short
    ox = x * BLOCK + (BLOCK / 2),
    oy = y * BLOCK + (BLOCK / 2);

  switch(Board[x][y])
  {
    case SPACE:
      return;
    case PIECE:
      w_pcircle(Window, ox, oy, SMALL);
      break;
    case EMPTY:
      w_circle(Window, ox, oy, LARGE);
      break;
  }
}

/* de/select a piece */
static void select_piece(short x, short y)
{
  x *= BLOCK;
  y *= BLOCK;
  w_box(Window, x, y, BLOCK, BLOCK);
}

/* moving a piece? -> select */
static int move_from(short x, short y)
{
  if(Board[x][y] == PIECE)
  {
    FromX = x; FromY = y;
    select_piece(x, y);
    return 1;
  }
  return 0;
}

/* checks if the move is valid and executes it */
static int move_to(short to_x, short to_y)
{
  short x, y, d1, d2;

  x  = (FromX + to_x) >> 1;		/* (from -> to) midpoint */
  y  = (FromY + to_y) >> 1;
  d1 = DIST(FromX, to_x);
  d2 = DIST(FromY, to_y);

  select_piece(FromX, FromY);		/* deselect */

  /* To empty, two block away, over a piece? */
  if(Board[to_x][to_y] == EMPTY &&
     ((d1 == 2 && d2 == 2) || (d1 == 2 && d2 == 0) || (d1 == 0 && d2 == 2)) &&
     Board[x][y] == PIECE)
  {
    draw_piece(FromX, FromY);		/* empty */
    Board[FromX][FromY] = EMPTY;

    draw_piece(x, y);
    Board[x][y] = EMPTY;

    Board[to_x][to_y] = PIECE;
    draw_piece(to_x, to_y);		/* fill with piece */

    return 1;
  }
  return 0;
}

/* ----------------------- */
static char Stack[PIECES*4];
static short Idx;

/* stack (FIFO) for undo */
static void push(short x, short y)
{
  Stack[Idx++] = FromY;
  Stack[Idx++] = FromX;
  Stack[Idx++] = y;
  Stack[Idx++] = x;
}

static int undo(void)
{
  short x1, y1, x2, y2;

  if(Idx < 4)
    return 0;

  /* remove moved-to */
  x1 = Stack[--Idx];
  y1 = Stack[--Idx];
  draw_piece(x1, y1);
  Board[x1][y1] = EMPTY;	/* remember that mode is invert... */

  /* insert moved-from */
  x2 = Stack[--Idx];
  y2 = Stack[--Idx];
  Board[x2][y2] = PIECE;
  draw_piece(x2, y2);

  /* insert removed-from-between */
  x2 = (x2 + x1) >> 1;
  y2 = (y2 + y1) >> 1;
  Board[x2][y2] = PIECE;
  draw_piece(x2, y2);

  return 1;
}

/* ----------------------- */
static int wait_input(void)
{
  WEVENT *ev;

  /* until no events, mouse button press */
  while((ev = w_queryevent(NULL, NULL, NULL, 0)) && ev->type == EVENT_MPRESS);

  /* until mouse button pressed or exited by a W event or with the ESC key */
  while(ev = w_queryevent(NULL, NULL, NULL, -1), ev->type != EVENT_MPRESS)
    if(((ev->type == EVENT_GADGET && ((ev->key == GADGET_EXIT) || (ev->key == GADGET_CLOSE)))) ||
       (ev->type == EVENT_KEY && (ev->key & 0xFF) == 'q'))
      return EXIT;

  if(ev->key == BUTTON_RIGHT)
    return UNDO;

  return PROCEED;
}
